### PR TITLE
cern: allow users based on 'IdentityClass'

### LINKED ISSUE #178: cern: lightweight account throws 500
[This change](https://github.com/inveniosoftware/invenio-oauthclient/pull/132) has introduced a bug for lightweight accounts. CERN lightweight accounts do not have have a `uidNumber`, so when logging in you get a stack trace on `external_id`:

`KeyError: 'uidNumber'`
```
invenio_oauthclient/contrib/cern.py in account_info at line 262
def account_info(remote, resp):
    """Retrieve remote account information used to find local user."""
    resource = get_resource(remote)
    email = resource['EmailAddress'][0]
    external_id = resource['uidNumber'][0]
```

<details><summary>Click to see an example of a login with a lightweight CERN account</summary>
<p>
```
[
  {
    "Type": "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier",
    "Value": "<email>@gmail.com"
  },
  {
    "Type": "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/upn",
    "Value": "<email>@gmail.com"
  },
  {
    "Type": "http://schemas.xmlsoap.org/claims/UPN",
    "Value": "<email>@gmail.com"
  },
  {
    "Type": "http://schemas.xmlsoap.org/claims/EmailAddress",
    "Value": "<email>@gmail.com"
  },
  {
    "Type": "http://schemas.xmlsoap.org/claims/CommonName",
    "Value": "j34hwtjhtg54g64j646g4"
  },
  {
    "Type": "http://schemas.microsoft.com/ws/2008/06/identity/claims/role",
    "Value": "CERN Externals"
  },
  {
    "Type": "http://schemas.xmlsoap.org/claims/DisplayName",
    "Value": "<email>@gmail.com"
  },
  {
    "Type": "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name",
    "Value": "<email>@gmail.com"
  },
  { "Type": "http://schemas.xmlsoap.org/claims/PersonID", "Value": "-3243534" },
  {
    "Type": "http://schemas.xmlsoap.org/claims/IdentityClass",
    "Value": "Unverified External"
  },
  { "Type": "http://schemas.xmlsoap.org/claims/Federation", "Value": "CERN" },
  { "Type": "http://schemas.xmlsoap.org/claims/AuthLevel", "Value": "Normal" },
  {
    "Type": "http://schemas.microsoft.com/ws/2008/06/identity/claims/authenticationmethod",
    "Value": "http://schemas.microsoft.com/ws/2008/06/identity/authenticationmethod/password"
  },
  {
    "Type": "http://schemas.microsoft.com/ws/2008/06/identity/claims/authenticationinstant",
    "Value": "2018-10-19T13:53:34.834Z"
  },
  { "Type": "urn:oauth:scope", "Value": "all" },
  {
    "Type": "http://schemas.xmlsoap.org/claims/Group",
    "Value": "All Exchange People"
  }
]
```
</p>
</details>

<details><summary>Click to see an example of a login with a normal CERN account</summary>
<p>
```
[
  {
    "Type": "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier",
    "Value": "<email>@cern.ch"
  },
  {
    "Type": "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/upn",
    "Value": "<email>@cern.ch"
  },
  {
    "Type": "http://schemas.xmlsoap.org/claims/UPN",
    "Value": "<email>@cern.ch"
  },
  {
    "Type": "http://schemas.xmlsoap.org/claims/EmailAddress",
    "Value": "<email>@cern.ch"
  },
  {
    "Type": "http://schemas.xmlsoap.org/claims/CommonName",
    "Value": "<username>"
  },
  {
    "Type": "http://schemas.microsoft.com/ws/2008/06/identity/claims/role",
    "Value": "CERN Users"
  },
  {
    "Type": "http://schemas.xmlsoap.org/claims/DisplayName",
    "Value": "<firstname lastname>"
  },
  {
    "Type": "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name",
    "Value": "<firstname lastname>"
  },
  { "Type": "http://schemas.xmlsoap.org/claims/Building", "Value": "<office number>" },
  { "Type": "http://schemas.xmlsoap.org/claims/Firstname", "Value": "<firstname>" },
  { "Type": "http://schemas.xmlsoap.org/claims/Lastname", "Value": "<lastname>" },
  { "Type": "http://schemas.xmlsoap.org/claims/Department", "Value": "<department>" },
  {
    "Type": "http://schemas.xmlsoap.org/claims/HomeInstitute",
    "Value": "CERN"
  },
  { "Type": "http://schemas.xmlsoap.org/claims/PersonID", "Value": "<personID>" },
  { "Type": "http://schemas.xmlsoap.org/claims/uidNumber", "Value": "<uidNumber>" },
  { "Type": "http://schemas.xmlsoap.org/claims/gidNumber", "Value": "<gidNumber>" },
  {
    "Type": "http://schemas.xmlsoap.org/claims/PreferredLanguage",
    "Value": "EN"
  },
  {
    "Type": "http://schemas.xmlsoap.org/claims/IdentityClass",
    "Value": "CERN Registered"
  },
  { "Type": "http://schemas.xmlsoap.org/claims/Federation", "Value": "CERN" },
  { "Type": "http://schemas.xmlsoap.org/claims/AuthLevel", "Value": "Normal" },
  {
    "Type": "http://schemas.microsoft.com/ws/2008/06/identity/claims/authenticationmethod",
    "Value": "http://schemas.microsoft.com/ws/2008/06/identity/authenticationmethod/password"
  },
  {
    "Type": "http://schemas.microsoft.com/ws/2008/06/identity/claims/authenticationinstant",
    "Value": "2018-10-19T12:41:25.214Z"
  },
  { "Type": "urn:oauth:scope", "Value": "all" },
  {
    "Type": "http://schemas.xmlsoap.org/claims/Group",
    "Value": "All Exchange People"
  },
  ... List of e-groups here ...
]
```
</p>
</details>

This is causing a 500 Internal Server Error.

### PR DESCRIPTION
**Pull request by zzacharo**
closes #178 .

### CODE REVIEW

**Comment by zzacharo** [comment]
Lightweight accounts do not have a `uidNumber` so we get a 500 error. I use the `PersonID` field as they have a negative id in that case.

**Comment by lnielsen** [comment]
The external id is treated as a string so probably not an issue as long as the two doesn't conflict. Did you try it out?

**Comment by ntarocco** [comment]
@lnielsen it appears that all of them are treated as string, no? I would go for string to be safe, and then let who is using that value do the parsing and eventually fail in a controlled manner.

@zzacharo I would also add, if the @lnielsen agrees, a new config variable for `cern` contrib, something like `OAUTHCLIENT_CERN_ALLOWED_IDENTITY_CLASS` = ['CERN Registered', 'CERN Shared']` (primary and secondary accounts, and probably also add service accounts by default, but I don't know the identity class for that) and fail login if not in the list.
With this, we can block lightweight account if needed (probably my most of the websites).
Probably you need to throw an exception in `account_info`, to be checked.

**Comment by lnielsen** [comment]
What was the conclusion in the end - that we close or integrate this PR?

**Comment by zzacharo** [comment]
@lnielsen @ntarocco @pamfilos I upgraded the PR with a new solution. We enable the cern accounts that belong to a specific `IdentityClass` so in that way be default we disable the lightweight accounts. Please review again.

File: invenio_oauthclient/errors.py [hunk 1]
```diff
@@ -71,3 +71,7 @@ def __init__(self, message, remote, response):
         super(OAuthClientError, self).__init__(
             self.description or message, remote, response
         )
+
+
+class OauthCernRejectedAccountError(OAuthResponseError):
```
Comment (lnielsen) [inline invenio_oauthclient/errors.py]: ``OAuthCERNRejectedAccountError`` (sorry to be annoying with the naming)

File: invenio_oauthclient/handlers.py [hunk 2]
```diff
@@ -227,6 +228,11 @@ def inner(*args, **kwargs):
             return oauth2_handle_error(
                 e.remote, e.response, e.code, e.uri, e.description
             )
+        except OauthCernRejectedAccountError as e:
+            current_app.logger.warning(e.message, exc_info=True)
+            flash(_('Cern account not allowed.'),
```
Comment (lnielsen) [inline invenio_oauthclient/handlers.py]: Cern -> CERN

**Comment by zzacharo** [comment]
@tiborsimko can you please check this PR so we can merge it if you don't see any problem?

**Pull request merged**